rfc/rfc-0000-input-output-correlation.md

```diff
@@ -0,0 +1,223 @@
+are evaluated and compared. If they are all equal, then the output is considered to be correlated with the input(s).
+
+When output cannot be correlated to input, Cartographer will not propagate the output forward. Doing so ensures we can
+always correlate outputs with inputs on subsequent resources, and also mitigates behavior of bad actors in the cluster
+as described
+in [Draft RFC Read Resources Only When In Success State](https://github.com/vmware-tanzu/cartographer/pull/556).
+
+In the absence of `correlationRules` Cartographer will continue to propagate outputs without attempting to correlate
+inputs to preserve existing functionality.
+
+For cases where the output does not contain necessary information to match against inputs, one may instead match against
+the spec of the resource:
+
+```yaml
+apiVersion: carto.run/v1alpha1
+kind: ClusterImageTemplate
+metadata:
+  name: example-image-building-thing
+spec:
+  template:
+    apiVersion: my-builder/v1
+    kind: ImageBuilderThatDoesntExposeInputDetailsInItsOutput
+    metadata:
+      name: ...
+    spec:
+      sourceUrl: "$(sources.source.url)$"
+      sourceRevision: "$(sources.source.url)$"
+correlationRules:
+  - input: "$(sources.source.url)"              #evaluated against inputs in template context
+    output: "$(spec.sourceUrl)"                 #evaluated against resource spec
+  - input: "$(sources.source.revision)"         #evaluated against inputs in template context
+    output: "$(spec.sourceRevision)"            #evaluated against resource spec
+```
+
+Since the spec of the resource may change independently of the output, when the `correlationRules` makes any assertion
+against the `spec`, Cartographer must also ensure that the current output is a result of processing the current spec by
+making additional assertions:
+
+* Resource is healthy, as per [Allow Resources to Report Status](https://github.com/vmware-tanzu/cartographer/pull/738)
+* Resource `observedGeneration == generation`
+
+If these assertions cannot be made, for example, if the resource does not implement the `observedGeneration` pattern,
+then Cartographer stops processing the resource and flags an error on it. Great care must also be taken that `healthy`
+status rules specified by the template author corresponds to the situation where the output is indeed a result of the
+spec with the current `observedGeneration`, otherwise outputs may not truly be the result of an input.
+
+Additionally, relying on the spec to make assertions imposes a further restriction to avoid stalling artifact
+propagation: Cartographer must not stamp out any further changes to the resource until it has finished processing the
+current spec completely. That is, it must have reached either a Healthy or Unhealthy status (as per
+[Allow Resources to Report Status](https://github.com/vmware-tanzu/cartographer/pull/738)).
```

Comment:
We need to avoid situations that could deadlock the supply chain. If an input is passed in that never resolves to Healthy True or False, the stamped resource is stuck, and can never be unstuck because we're no longer writing to its spec.

Reply:
I've added a `correlationTimeout` setting to the template for now; I'm not sure there's a better solution whilst remaining agnostic.